### PR TITLE
add maxpool2D, add more tests, handle integer parameters for maxpool2D

### PR DESCRIPTION
**Pull request by migeed-z**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61310
* #61265
* #61239
* #61238
* **#61188**

Differential Revision: [D29626303](https://our.internmc.facebook.com/intern/diff/D29626303)

### CODE REVIEW

File: torch/fx/experimental/graph_gradual_typechecker.py [hunk 1]
```diff
@@ -269,6 +283,50 @@ def relu_inference_rule(n: Node, op_type):
...
+        new_type_list[-1] = w_out
+        new_type_list[-2] = h_out
+        return TensorType(tuple(new_type_list))
+
+    else:
+        raise TypeError(f'Wrong size {typ} for {op_type}')
+
+
+@register_inference_rule(torch.nn.MaxPool2d)
+def maxpool2d_inference_rule(n: Node, op_type):
```
Comment (jamesr66a) [inline torch/fx/experimental/graph_gradual_typechecker.py:302]: same here, rename `op_type`

File: test/fx/test_gradual_type.py [hunk 2]
```diff
@@ -460,5 +460,57 @@ def forward(self, x: TensorType((2, 2, 4, 5))):
                 assert isinstance(n.type, TensorType)
                 assert torch.Size(n.type.__args__) == B.forward(torch.rand(2, 2, 4, 5)).size()
 
+    def test_type_check_conv2D_and_maxpool2d(self):
```
Comment (jamesr66a) [inline test/fx/test_gradual_type.py:463]: Same comment here as conv2d, if we can be more exhaustive about testing `stride`, `padding`, etc that would be good

**Review by jamesr66a — COMMENTED** [review]
Looks pretty good!

**Comment by migeed-z** [comment]
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29626303).

File: torch/fx/experimental/graph_gradual_typechecker.py [hunk 3]
```diff
@@ -277,6 +289,37 @@ def relu_inference_rule(n: Node, module_instance):
...
+
+@register_inference_rule(torch.nn.MaxPool2d)
+def maxpool2d_inference_rule(n: Node, module_instance):
+    """
+    Given a MaxPool2D instance and a node check the following conditions:
+    - Input size matches size 3 or 4
+    - Current node type is consistent with the output type we will calculate
+    - Input size matches output size and the last two dimensions of the output
+      are w_out and h_out. The remaining dimensions are the same as the input
+    - Our final result is the greatest lower bound of the output we calculate
```
Comment (jamesr66a) [inline torch/fx/experimental/graph_gradual_typechecker.py:316]: Is this supposed to say "upper bound"?

**Review by jamesr66a — APPROVED** [review]
LGTM. I think you said you were going to add more tests, but I'm just going to preemptively stamp since they'll probably look the same as the conv2d tests

**Comment by migeed-z** [comment]
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29626303).

**Comment by migeed-z** [comment]
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29626303).

**Comment by migeed-z** [comment]
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29626303).

**Comment by migeed-z** [comment]
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29626303).
